File: perl/test/closures3.pl
#!/usr/bin/perl -w

sub accum {
    my $curr = 0;
    sub {
        my $param = shift;
        $curr += $param;
    }
}


my $acc1 = accum();

print "a1 ",$acc1->(3),"\n";
print "a1 ",$acc1->(5),"\n";

my $acc2 = accum();

print "a2 ",$acc2->(1),"\n";

print "a1 ",$acc1->(2),"\n";


print "a2 ",$acc2->(7),"\n";

print "a1 ",$acc1->(10),"\n";

print "a2 ",$acc2->(4),"\n";

File: perl/test/closures3.java
interface Func {
    public int run(int param);
}

public class closures3 {

    public static Func accum(int start) {
        final int[] curr = {start};
        return new Func() {
                public int run(int param) {
                    curr[0] += param;
                    return curr[0];
                }
            };
    }

    public static void main(String[] args) {
        Func acc1 = accum(0);

        System.out.println("a1 "+acc1.run(3));
        System.out.println("a1 "+acc1.run(5));

        Func acc2 = accum(10);

        System.out.println("a2 "+acc2.run(1));

        System.out.println("a1 "+acc1.run(2));

        
        System.out.println("a2 "+acc2.run(7));
        
        System.out.println("a1 "+acc1.run(10));
        
        System.out.println("a2 "+acc2.run(4));
    }

}
